src/dimmwitted.cc

```diff
@@ -143,19 +162,108 @@ void DimmWitted::inference() {
 
     double elapsed = t.elapsed();
     if (should_show_progress) {
-      std::cout << "" << elapsed << " sec.";
-      std::cout << "," << (nvar * n_samplers_) / elapsed << " vars/sec"
-                << std::endl;
+      std::streamsize ss = std::cout.precision();
+      std::cout << std::setprecision(3) << "" << elapsed << " sec."
+                << "," << (nvar * n_samplers_) / elapsed << " vars/sec"
+                << std::endl
+                << std::setprecision(ss);
     }
   }
 
   double elapsed = t_total.elapsed();
   std::cout << "TOTAL INFERENCE TIME: " << elapsed << " sec." << std::endl;
 }
 
+template <typename T>
+void inspect_vector(std::vector<T> &vec) {
+  std::streamsize ss = std::cout.precision();
+  std::cout << std::setprecision(3)
+            << std::vector<float>(vec.begin(), vec.begin() + 3) << " ... "
+            << std::vector<float>(vec.end() - 3, vec.end())
+            << std::setprecision(ss) << std::endl;
+}
+
+bool DimmWitted::ps_update_weights(int epochs, std::vector<float> &delta) {
+  InferenceResult &infrs = samplers[0].infrs;
+  const size_t nweight = infrs.nweights;
+
+  // REQUEST FORMAT:
+  //   STRING worker_id
+  //   STRING msgtype
+  //   INT epochs
+  //   FLOAT32[]::BYTES grads
+
+  std::cout << "\tsending grads[" << delta.size() << "] ";
+  inspect_vector(delta);
+  msgpack::sbuffer sbuf;
+  msgpack::packer<msgpack::sbuffer> pk(&sbuf);
+  pk.pack(opts.worker_id);
+  pk.pack("GRADIENTS");
+  pk.pack(epochs);
+  pk.pack_bin(4 * delta.size());
+  pk.pack_bin_body(reinterpret_cast<char const *>(delta.data()),
+                   4 * delta.size());
+
+  // TODO: compression
+  zmq::message_t msg(sbuf.data(), sbuf.size(), NULL /* no de-allocate */);
+  ps_socket->send(msg);
+
+  // RESPONSE FORMAT:
+  //   STRING command
+  //   FLOAT32[]::BYTES weights
+
+  zmq::message_t reply;
+  ps_socket->recv(&reply);
+  msgpack::unpacker pac;
+  pac.reserve_buffer(reply.size());
+  memcpy(pac.buffer(), reply.data(), reply.size());
+  pac.buffer_consumed(reply.size());
+
+  std::string command;
+  msgpack::object_handle oh;
+  pac.next(oh);
+  oh.get().convert(command);
+  pac.next(oh);
+  std::vector<char> bytes;
+  oh.get().convert(bytes);
+
+  // HACK: abusing delta to store weights
+  memcpy(delta.data(), bytes.data(), bytes.size());
+  std::cout << "\treceived wgts[" << delta.size() << "] ";
+  inspect_vector(delta);
```

Comment:
Sure.  Then let's put a common prefix for these interactions with the parameter server.  And also a `std::flush` at the end?